### ISSUE TITLE
Generated client use

### ISSUE THREAD

**Issue by rookie7799** [go-swagger/go-swagger#1544]
## Problem statement
Is there a reason why a Payload gets returned only when "Accepted"? 
An API returns useful messages when something is off and as far as I can tell I can only get those messages from "Error()" which is very inconvenient. 

Why not return a Payload struct along with err so one can use it if need be?

Unless off course I'm completely miss-understanding the "client" usage.

## Swagger specification
2.0

## Steps to reproduce
1. Generate spec
2. Generate client

## Environment
swagger version: 0.13.0
go version: 1.9.2
OS:   CentOS 6

**Comment by casualjim** [comment]
You cna just cast to the underlying error and get the payload

The error is just a common interface all the non 2xx status codes implement. So you can distinguish between success and error by method result (left, right)
And then you can use a type switch to get to the exact error,

**Comment by rookie7799** [comment]
Sorry for my lack of knowledge here, I have this snippet of code:
```
ulp := puser.NewUserLoginParams()
pl := models.PostLogin{Email: c.PostForm("email"), Pwd: c.PostForm("pwd")}
ulp.SetBody(&pl)
ula, err := client.User.UserLogin(ulp)
```

How would I get the payload info from `err`? I mean wouldn't I need to know which response to "type assert" to ?

Thank you for your help!

**Comment by rookie7799** [comment]
This is the fun that was generated by "swagger generate client"
```
/*
UserLogin User logging in
*/
func (a *Client) UserLogin(params *UserLoginParams) (*UserLoginAccepted, error) {
	// TODO: Validate the params before sending
	if params == nil {
		params = NewUserLoginParams()
	}

	result, err := a.transport.Submit(&runtime.ClientOperation{
		ID:                 "user_login",
		Method:             "POST",
		PathPattern:        "/user/login",
		ProducesMediaTypes: []string{"application/json"},
		ConsumesMediaTypes: []string{"application/json"},
		Schemes:            []string{"http", "https"},
		Params:             params,
		Reader:             &UserLoginReader{formats: a.formats},
		Context:            params.Context,
		Client:             params.HTTPClient,
	})
	if err != nil {
		return nil, err
	}
	return result.(*UserLoginAccepted), nil
}
```

Not sure if its even possible to get an actual Payload data from `err`.

**Comment by casualjim** [comment]
it will be another type like UserLoginDefault  that you can cast err to. The generator generates a struct for each response that's mentioned in the swagger file

**Comment by kylej-ibm** [comment]
Was this ever solved?  I have the same issue. I need the payload that was originally returned (in our case its a model.Error object). However the Error() function always gets called for request and obfuscates that object with an string...  I dont understand how you can "cast" err when using the client...

Generated Go Client
```go
func (o *FooEntity) Error() string {
	return fmt.Sprintf("[POST /foo][%d] PostUnprocessableEntity  %+v", 422, *o.Payload)
}

func (o *FooEntity) readResponse(response runtime.ClientResponse, consumer runtime.Consumer, formats strfmt.Registry) error {

	o.Payload = new(models.Error)

	// response payload
	if err := consumer.Consume(response.Body(), o.Payload); err != nil && err != io.EOF {
		return err
	}

	return nil
}
```

Client Code
```
ok, err := client.FooSomething.FooPost(params, clientAuth)
```

err is always the string "[POST /foo][422] PostUnprocessableEntity  &{Description: Error:}" and my contents of models.Error are missing...

**Comment by nightlyone** [comment]
So the generator generates more something like an example usage client for the happy path and not a fully usable one that actually reports a wrong http status as an error code.

Am I understanding this correctly? 

Would you @casualjim  accept some extension to mark a response as an error condition and return it as such?

I believe supporting error responses will make the generated client much more useful as endpoints then still get added/deleted/renamed as we change the swagger.json/swagger.yaml and we can directly use them in higher level application code.

**Comment by fredbi** [comment]
@nightlyone 
> So the generator generates more something like an example usage client for the happy path and not a fully usable one that actually reports a wrong http status as an error code.

> Am I understanding this correctly?

No. The generated client is actually fully usable and returns an error for any error status code.
Error responses may have a body, provided this is defined in the spec.

The only point that might be confusing, but looks to me as pretty common in go is the need to proceed with error type assertion to convert back the `error` type in a useful structure.

**Comment by fredbi** [comment]
@rookie7799 @nightlyone you'll find here an example showing how to assert error types.
If this error type has been specified with some body, you'll be able to retrieve this as a payload from the error.
https://github.com/go-openapi/runtime/issues/121

**Comment by nightlyone** [comment]
Thank you very much @fredbi , this is pretty useful!

What about generating the necessary type switch on the error message as an example in the API client for each function? That will help getting people started quickly. It will also document all the expected responses.

Would such a contribution be welcomed?
